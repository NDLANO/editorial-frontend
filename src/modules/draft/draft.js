/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { handleActions, createAction } from 'redux-actions';
import { createSelector } from 'reselect';
import { getLocale } from '../locale/locale';
import { convertFieldWithFallback } from '../../util/convertFieldWithFallback';
import formatDate from '../../util/formatDate';

export const fetchDraft = createAction('FETCH_DRAFT');
export const setDraft = createAction('SET_DRAFT');
export const updateDraft = createAction('UPDATE_DRAFT');
export const updateDraftSuccess = createAction('UPDATE_DRAFT_SUCCESS');
export const updateDraftError = createAction('UPDATE_DRAFT_ERROR');
export const updateStatusDraft = createAction('UPDATE_STATUS_DRAFT');

export const actions = {
  updateDraft,
  fetchDraft,
  setDraft,
  updateDraftSuccess,
  updateDraftError,
  updateStatusDraft,
};

const initalState = {
  all: {},
  isSaving: false,
};

export default handleActions(
  {
    [setDraft]: {
      next: (state, action) => ({
        ...state,
        all: { ...state.all, [action.payload.id]: { ...action.payload } },
      }),
    },
    [updateDraft]: {
      next: state => ({
        ...state,
        isSaving: true,
      }),
    },
    [updateStatusDraft]: {
      next: state => ({
        ...state,
        isSaving: true,
      }),
    },
    [updateDraftSuccess]: {
      next: state => ({
        ...state,
        isSaving: false,
      }),
    },
    [updateDraftError]: {
      next: state => ({
        ...state,
        isSaving: false,
      }),
    },
  },
  initalState,
);

const getDraftsFromState = state => state.drafts;

export const getDraftById = draftId =>
  createSelector([getDraftsFromState], drafts => drafts.all[draftId]);

export const getSaving = createSelector(
  [getDraftsFromState],
  drafts => drafts.isSaving,
);

const getLanguageFromField = (object, field, fallback = undefined) =>
  object && object[field] ? object[field].language : fallback;

export const articleConverter = (article, locale) => ({
  ...article,
  title: convertFieldWithFallback(article, 'title', ''),
  introduction: convertFieldWithFallback(article, 'introduction', ''),
  visualElement: convertFieldWithFallback(article, 'visualElement', {}),
  content: convertFieldWithFallback(article, 'content', ''),
  footnotes:
    article.content && article.content.footNotes
      ? article.content.footNotes
      : undefined,
  metaDescription: convertFieldWithFallback(article, 'metaDescription', ''),
  tags: convertFieldWithFallback(article, 'tags', []),
  language:
    article && article.tags
      ? getLanguageFromField(article, 'tags')
      : getLanguageFromField(article, 'content', locale),
  created: formatDate(article.created, locale),
  updated: formatDate(article.updated, locale),
});

export const getDraft = articleId =>
  createSelector(
    [getDraftById(articleId), getLocale],
    (article, locale) =>
      article ? articleConverter(article, locale) : undefined,
  );
