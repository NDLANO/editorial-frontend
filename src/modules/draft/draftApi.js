/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
import queryString from 'query-string';
import {
  resolveJsonOrRejectWithError,
  apiResourceUrl,
  fetchAuthorized,
} from '../../util/apiHelpers';

const baseUrl = apiResourceUrl('/draft-api/v1/drafts');
const baseAgreementsUrl = apiResourceUrl('/draft-api/v1/agreements');
const baseFileUrl = apiResourceUrl('/draft-api/v1/files');

export const fetchTags = language => {
  const query = queryString.stringify({ size: 7000, language });
  return fetchAuthorized(`${baseUrl}/tags/?${query}`).then(
    resolveJsonOrRejectWithError,
  );
};

export const fetchLicenses = () =>
  fetchAuthorized(`${baseUrl}/licenses/`).then(resolveJsonOrRejectWithError);

export const fetchDraft = (id, language) => {
  const query = queryString.stringify({ language });
  const url = language
    ? `${baseUrl}/${id}?${query}&fallback=true`
    : `${baseUrl}/${id}`;
  return fetchAuthorized(url).then(resolveJsonOrRejectWithError);
};

export const cloneDraft = (id, language) => {
  const query = queryString.stringify({ language });
  const url = language
    ? `${baseUrl}/clone/${id}?${query}&fallback=true`
    : `${baseUrl}/clone/${id}`;
  return fetchAuthorized(url, { method: 'POST' }).then(
    resolveJsonOrRejectWithError,
  );
};

export const updateDraft = draft =>
  fetchAuthorized(`${baseUrl}/${draft.id}`, {
    method: 'PATCH',
    body: JSON.stringify(draft),
  }).then(resolveJsonOrRejectWithError);

export const createDraft = draft =>
  fetchAuthorized(`${baseUrl}/`, {
    method: 'POST',
    body: JSON.stringify(draft),
  }).then(resolveJsonOrRejectWithError);

export const fetchDraftHistory = (id, language) => {
  const query = queryString.stringify({ language });
  const url = language
    ? `${baseUrl}/${id}/history?${query}&fallback=true`
    : `${baseUrl}/${id}/history`;
  return fetchAuthorized(url).then(resolveJsonOrRejectWithError);
};

export const deleteLanguageVersion = (id, language) =>
  fetchAuthorized(`${baseUrl}/${id}/language/${language}`, {
    method: 'DELETE',
  }).then(resolveJsonOrRejectWithError);

export const fetchNewArticleId = id => {
  const url = `${baseUrl}/external_id/${id}`;
  return fetchAuthorized(url).then(resolveJsonOrRejectWithError);
};

export const validateDraft = (id, draft) =>
  fetchAuthorized(`${baseUrl}/${id}/validate/`, {
    method: 'PUT',
    body: JSON.stringify(draft),
  }).then(resolveJsonOrRejectWithError);

export const updateStatusDraft = (id, status) =>
  fetchAuthorized(`${baseUrl}/${id}/status/${status}`, {
    method: 'PUT',
  }).then(resolveJsonOrRejectWithError);

export const fetchAgreements = query =>
  fetchAuthorized(`${baseAgreementsUrl}?query=${query}`).then(
    resolveJsonOrRejectWithError,
  );

export const fetchAgreement = id =>
  fetchAuthorized(`${baseAgreementsUrl}/${id}`).then(
    resolveJsonOrRejectWithError,
  );

export const updateAgreement = agreement =>
  fetchAuthorized(`${baseAgreementsUrl}/${agreement.id}`, {
    method: 'PATCH',
    body: JSON.stringify(agreement),
  }).then(resolveJsonOrRejectWithError);

export const createAgreement = agreement =>
  fetchAuthorized(`${baseAgreementsUrl}/`, {
    method: 'POST',
    body: JSON.stringify(agreement),
  }).then(resolveJsonOrRejectWithError);

export const fetchStatusStateMachine = () =>
  fetchAuthorized(`${baseUrl}/status-state-machine/`).then(
    resolveJsonOrRejectWithError,
  );

export const uploadFile = formData =>
  fetchAuthorized(`${baseFileUrl}/`, {
    method: 'POST',
    headers: { 'Content-Type': undefined },
    body: formData,
  }).then(resolveJsonOrRejectWithError);

export const searchDrafts = query =>
  fetchAuthorized(`${baseUrl}/search/`, {
    method: 'POST',
    body: JSON.stringify(query),
  }).then(resolveJsonOrRejectWithError);
