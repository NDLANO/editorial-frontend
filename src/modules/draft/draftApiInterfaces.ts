/**
 * Copyright (c) 2021-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

interface Author {
  type: string;
  name: string;
}

interface Status {
  status: string;
  other: string[];
}

export interface DraftStatusStateMachineType {
  QUALITY_ASSURED: string[];
  PUBLISHED: string[];
  QUEUED_FOR_LANGUAGE: string[];
  AWAITING_UNPUBLISHING: string[];
  QUALITY_ASSURED_DELAYED: string[];
  QUEUED_FOR_PUBLISHING: string[];
  QUEUED_FOR_PUBLISHING_DELAYED: string[];
  IMPORTED: string[];
  DRAFT: string[];
  USER_TEST: string[];
  PROPOSAL: string[];
  TRANSLATED: string[];
  UNPUBLISHED: string[];
  AWAITING_QUALITY_ASSURANCE: string[];
  ARCHIVED: string[];
}

export interface UploadedFileType {
  filename: string;
  mime: string;
  extension: string;
  path: string;
}

export interface LicenseResult {
  license: string;
  description?: string;
  url?: string;
}

export interface TagSearchResult {
  totalCount: number;
  page: number;
  pageSize: number;
  language: string;
  results: string[];
}

export interface TagResult {
  tags: string[];
  language: string;
}

export interface GrepCodesSearchResult {
  totalCount: number;
  page: number;
  pageSize: number;
  results: string[];
}

export interface AgreementSearchResult {
  totalCount: number;
  page?: number;
  pageSize: number;
  language: string;
  results: AgreementSummary[];
}

interface AgreementSummary {
  id: number;
  title: string;
  license: string;
}

export type DraftStatusTypes =
  | 'IMPORTED'
  | 'DRAFT'
  | 'PUBLISHED'
  | 'PROPOSAL'
  | 'QUEUED_FOR_PUBLISHING'
  | 'USER_TEST'
  | 'AWAITING_QUALITY_ASSURANCE'
  | 'QUEUED_FOR_LANGUAGE'
  | 'TRANSLATED'
  | 'QUALITY_ASSURED'
  | 'QUALITY_ASSURED_DELAYED'
  | 'QUEUED_FOR_PUBLISHING_DELAYED'
  | 'AWAITING_UNPUBLISHING'
  | 'UNPUBLISHED'
  | 'ARCHIVED';

export interface DraftSearchQuery {
  query?: string;
  language?: string;
  license?: string;
  page?: number;
  pageSize?: number;
  idList?: number[];
  articleTypes?: string[];
  sort?: string;
  scrollId?: string;
  fallback?: boolean;
  grepCodes?: string[];
}

export interface DraftSearchResult {
  totalCount: number;
  page?: number;
  pageSize: number;
  results: DraftSearchSummary[];
}

export interface DraftSearchSummary {
  id: number;
  title: {
    title: string;
    language: string;
  };
  visualElement?: {
    visualElement: string;
    language: string;
  };
  introduction?: {
    introduction: string;
    language: string;
  };
  url: string;
  license: string;
  articleType: string;
  supportedLanguages: string[];
  tags?: {
    tags: string[];
    language: string;
  };
  notes: string[];
  users: string[];
  grepCodes: string[];
}

export interface UserDataApiType {
  userId: string;
  savedSearches?: string[];
  latestEditedArticles?: string[];
  favoriteSubjects?: string[];
}

export interface UpdatedUserDataApiType {
  savedSearches?: string[];
  latestEditedArticles?: string[];
  favoriteSubjects?: string[];
}

export interface AgreementApiType {
  id: number;
  title: string;
  content: string;
  copyright: Copyright;
  created: string;
  updated: string;
  updatedBy: string;
}

export interface UpdatedAgreementApiType {
  id?: number; // Discarded by backend, used only by endpoint
  title?: string;
  content?: string;
  copyright?: Copyright;
}

export interface NewAgreementApiType {
  title: string;
  content: string;
  copyright: Copyright;
}

export interface DraftApiType {
  id: number;
  oldNdlaUrl?: string;
  revision: number;
  status: Status;
  title?: {
    title: string;
    language: string;
  };
  content?: {
    content: string;
    language: string;
  };
  copyright?: Copyright;
  tags?: {
    tags: string[];
    language: string;
  };
  requiredLibraries: {
    mediaType: string;
    name: string;
    url: string;
  }[];
  visualElement?: {
    visualElement: string;
    language: string;
  };
  introduction?: {
    introduction: string;
    language: string;
  };
  metaDescription?: {
    metaDescription: string;
    language: string;
  };
  metaImage?: {
    url: string;
    alt: string;
    language: string;
  };
  created: string;
  updated: string;
  updatedBy: string;
  published: string;
  articleType: string;
  supportedLanguages: string[];
  notes: {
    note: string;
    user: string;
    status: Status;
    timestamp: string;
  }[];
  editorLabels: string[];
  grepCodes: string[];
  conceptIds: number[];
  availability: string;
}

export interface UpdatedDraftApiType {
  id?: number; // Discarded by backend, used only by endpoint
  revision: number;
  language?: string;
  title?: string;
  status?: string;
  published?: string;
  content?: string;
  tags?: string[];
  introduction?: string;
  metaDescription?: string;
  // null indicates delete, undefined from optional indicates no change.
  // if we want to delete given metaImage, send null as a value.
  metaImage?: {
    id: string;
    alt: string;
  } | null;
  visualElement?: string;
  copyright?: Copyright;
  requiredLibraries?: {
    mediaType: string;
    name: string;
    url: string;
  }[];
  articleType?: string;
  notes?: string[];
  editorLabels?: string[];
  grepCodes?: string[];
  conceptIds?: number[];
  createNewVersion?: boolean;
  availability?: string;
}

export interface NewDraftApiType {
  language: string;
  title: string;
  published?: string;
  content?: string;
  tags: string[];
  introduction?: string;
  metaDescription?: string;
  metaImage?: {
    url: string;
    alt: string;
  };
  visualElement?: string;
  copyright?: Copyright;
  requiredLibraries: {
    mediaType: string;
    name: string;
    url: string;
  }[];
  articleType: string;
  notes: string[];
  editorLabels: string[];
  grepCodes: string[];
  conceptIds: number[];
  availability?: string;
}

interface Copyright {
  license?: {
    license: string;
    description?: string;
    url?: string;
  };
  origin?: string;
  creators: Author[];
  processors: Author[];
  rightsholders: Author[];
  agreementId?: number;
  validFrom?: string;
  validTo?: string;
}
