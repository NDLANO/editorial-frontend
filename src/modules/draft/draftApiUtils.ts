/**
 * Copyright (c) 2021-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { ILicense as ImageApiLicense } from '@ndla/types-image-api';
import { ILicense as DraftApiLicense } from '@ndla/types-draft-api';

export const draftLicensesToImageLicenses = (licenses: DraftApiLicense[]): ImageApiLicense[] =>
  licenses.map((l) => ({ license: l.license, description: l.description ?? '', url: l.url }));
