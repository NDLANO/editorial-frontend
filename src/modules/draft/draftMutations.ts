/**
 * Copyright (c) 2022-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { IArticle, IUpdatedArticle } from '@ndla/types-backend/draft-api';
import { useMutation, UseMutationOptions } from '@tanstack/react-query';
import { copyRevisionDates, updateDraft } from './draftApi';

export const useUpdateDraftMutation = (
  options?: Partial<UseMutationOptions<IArticle, unknown, { id: number; body: IUpdatedArticle }>>,
) => {
  return useMutation<IArticle, undefined, { id: number; body: IUpdatedArticle }>({
    mutationFn: (vars) => updateDraft(vars.id, vars.body),
    ...options,
  });
};

export const useCopyRevisionDates = (
  options?: UseMutationOptions<void, unknown, { nodeId: string }>,
) => {
  return useMutation<void, unknown, { nodeId: string }>({
    mutationFn: (vars) => copyRevisionDates(vars.nodeId),
    ...options,
  });
};
