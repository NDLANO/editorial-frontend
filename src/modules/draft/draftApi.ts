/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
import queryString from 'query-string';
import {
  ILicense,
  INewArticle,
  IUpdatedArticle,
  IArticle,
  ITagsSearchResult,
  IArticleTag,
  IGrepCodesSearchResult,
  IUserData,
  ISearchResult,
  IUpdatedUserData,
  IUploadedFile,
} from '@ndla/types-backend/draft-api';
import {
  resolveJsonOrRejectWithError,
  apiResourceUrl,
  fetchAuthorized,
} from '../../util/apiHelpers';
import { DraftSearchQuery } from './draftApiInterfaces';
import { resolveVoidOrRejectWithError } from '../../util/resolveJsonOrRejectWithError';
import { DraftStatusType, DraftStatusStateMachineType } from '../../interfaces';

const baseUrl: string = apiResourceUrl('/draft-api/v1/drafts');
const baseAgreementsUrl: string = apiResourceUrl('/draft-api/v1/agreements');
const baseFileUrl: string = apiResourceUrl('/draft-api/v1/files');
const baseUserDataUrl: string = apiResourceUrl('/draft-api/v1/user-data');

export const fetchDraft = async (id: number | string, language?: string): Promise<IArticle> => {
  const query = queryString.stringify({ language });
  const url = language ? `${baseUrl}/${id}?${query}&fallback=true` : `${baseUrl}/${id}`;
  return fetchAuthorized(url).then((r) => resolveJsonOrRejectWithError<IArticle>(r));
};

export const fetchDrafts = async (ids: number[], language?: string): Promise<IArticle[]> => {
  const query = queryString.stringify({
    ids: ids.join(','),
    language,
    fallback: true,
    page: 1,
    'page-size': ids.length,
  });
  return fetchAuthorized(`${baseUrl}/ids/?${query}`, {
    method: 'GET',
  }).then((r) => resolveJsonOrRejectWithError<[IArticle]>(r));
};

export const updateDraft = async (
  id: number,
  draft: IUpdatedArticle,
  versionHash = 'default',
): Promise<IArticle> =>
  fetchAuthorized(`${baseUrl}/${id}`, {
    method: 'PATCH',
    headers: { VersionHash: versionHash },
    body: JSON.stringify(draft),
  }).then((r) => resolveJsonOrRejectWithError<IArticle>(r));

export const createDraft = async (draft: INewArticle): Promise<IArticle> =>
  fetchAuthorized(`${baseUrl}/`, {
    method: 'POST',
    body: JSON.stringify(draft),
  }).then((r) => resolveJsonOrRejectWithError<IArticle>(r));

export const searchDrafts = async (query: DraftSearchQuery): Promise<ISearchResult> =>
  fetchAuthorized(`${baseUrl}/search/`, {
    method: 'POST',
    body: JSON.stringify(query),
  }).then((r) => resolveJsonOrRejectWithError<ISearchResult>(r));

export const searchAllDrafts = async (
  ids: number[],
  language?: string,
  sort?: string,
): Promise<ISearchResult> => {
  const query = queryString.stringify({
    ids: ids.join(','),
    language,
    page: 1,
    'page-size': ids.length,
    sort,
    fallback: true,
  });

  return fetchAuthorized(`${baseUrl}/?${query}`, {
    method: 'GET',
  }).then((r) => resolveJsonOrRejectWithError<ISearchResult>(r));
};

export const cloneDraft = async (
  id: number,
  language?: string,
  addCopyPostfixToArticleTitle: boolean = true,
): Promise<IArticle> => {
  const query = queryString.stringify({
    language,
    'copied-title-postfix': addCopyPostfixToArticleTitle,
    fallback: true,
  });
  const url = `${baseUrl}/clone/${id}?${query}`;
  return fetchAuthorized(url, { method: 'POST' }).then((r) =>
    resolveJsonOrRejectWithError<IArticle>(r),
  );
};

export const fetchDraftHistory = async (id: number, language?: string): Promise<IArticle[]> => {
  const query = queryString.stringify({ language });
  const url = language
    ? `${baseUrl}/${id}/history?${query}&fallback=true`
    : `${baseUrl}/${id}/history`;
  return fetchAuthorized(url).then((r) => resolveJsonOrRejectWithError<IArticle[]>(r));
};

export const deleteLanguageVersion = async (id: number, language: string): Promise<IArticle> =>
  fetchAuthorized(`${baseUrl}/${id}/language/${language}`, {
    method: 'DELETE',
  }).then((r) => resolveJsonOrRejectWithError<IArticle>(r));

export const fetchNewArticleId = async (id: number): Promise<{ id: number }> => {
  const url = `${baseUrl}/external_id/${id}`;
  return fetchAuthorized(url).then((r) => resolveJsonOrRejectWithError<{ id: number }>(r));
};

export const validateDraft = async (id: number, draft: IUpdatedArticle): Promise<{ id: number }> =>
  fetchAuthorized(`${baseUrl}/${id}/validate/`, {
    method: 'PUT',
    body: JSON.stringify(draft),
  }).then((r) => resolveJsonOrRejectWithError<{ id: number }>(r));

export const updateStatusDraft = async (id: number, status: DraftStatusType): Promise<IArticle> =>
  fetchAuthorized(`${baseUrl}/${id}/status/${status}`, {
    method: 'PUT',
  }).then((r) => resolveJsonOrRejectWithError<IArticle>(r));

export const fetchTags = async (language: string): Promise<IArticleTag> => {
  const query = queryString.stringify({ size: 7000, language });
  return fetchAuthorized(`${baseUrl}/tags/?${query}`).then((r) =>
    resolveJsonOrRejectWithError<IArticleTag>(r),
  );
};

export const fetchSearchTags = async (
  input: string,
  language: string,
): Promise<ITagsSearchResult> =>
  fetchAuthorized(`${baseUrl}/tag-search/?language=${language}&query=${input}&fallback=true`).then(
    (r) => resolveJsonOrRejectWithError<ITagsSearchResult>(r),
  );

export const fetchLicenses = async (): Promise<ILicense[]> =>
  fetchAuthorized(`${baseUrl}/licenses/`).then((r) => resolveJsonOrRejectWithError<ILicense[]>(r));

export const fetchGrepCodes = async (query: string): Promise<IGrepCodesSearchResult> =>
  fetchAuthorized(`${baseUrl}/grep-codes/?query=${query}`).then((r) =>
    resolveJsonOrRejectWithError<IGrepCodesSearchResult>(r),
  );

export const fetchUserData = async (): Promise<IUserData> =>
  fetchAuthorized(`${baseUserDataUrl}`).then((r) => resolveJsonOrRejectWithError<IUserData>(r));

export const updateUserData = async (userData: IUpdatedUserData): Promise<IUserData> =>
  fetchAuthorized(`${baseUserDataUrl}`, {
    method: 'PATCH',
    body: JSON.stringify(userData),
  }).then((r) => resolveJsonOrRejectWithError<IUserData>(r));

export const fetchStatusStateMachine = async (
  id?: number,
): Promise<DraftStatusStateMachineType> => {
  const idParam = id ? `?articleId=${id}` : '';
  return fetchAuthorized(`${baseUrl}/status-state-machine/${idParam}`).then((r) =>
    resolveJsonOrRejectWithError<DraftStatusStateMachineType>(r),
  );
};

export const copyRevisionDates = (nodeId: string): Promise<void> => {
  return fetchAuthorized(`${baseUrl}/copyRevisionDates/${nodeId}`, {
    method: 'POST',
  }).then((r) => resolveVoidOrRejectWithError(r));
};

export const headFileAtRemote = async (fileUrl: string): Promise<boolean> => {
  const res = await fetch(fileUrl, {
    method: 'HEAD',
  });
  return res.status === 200;
};

export const uploadFile = async (formData: any): Promise<IUploadedFile> =>
  fetchAuthorized(`${baseFileUrl}/`, {
    method: 'POST',
    headers: { 'Content-Type': undefined },
    body: formData,
  }).then((r) => resolveJsonOrRejectWithError<IUploadedFile>(r));

export const deleteFile = async (fileUrl: string): Promise<void> => {
  const query = encodeURIComponent(fileUrl);
  fetchAuthorized(`${baseFileUrl}/?path=${query}`, {
    method: 'DELETE',
  }).then(resolveVoidOrRejectWithError);
};
