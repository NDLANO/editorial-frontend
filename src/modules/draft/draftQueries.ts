/**
 * Copyright (c) 2021-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { useMutation, useQuery, useQueryClient, UseQueryOptions } from "@tanstack/react-query";
import {
  ILicense,
  IArticle,
  IUserData,
  IUpdatedUserData,
  ISearchResult,
  ITagsSearchResult,
} from "@ndla/types-backend/draft-api";
import {
  fetchDraft,
  fetchLicenses,
  fetchStatusStateMachine,
  fetchUserData,
  updateUserData,
  searchAllDrafts,
  fetchDraftHistory,
  fetchSearchTags,
} from "./draftApi";
import { DraftSearchQuery } from "./draftApiInterfaces";
import { DraftStatusStateMachineType } from "../../interfaces";
import {
  DRAFT,
  DRAFT_STATUS_STATE_MACHINE,
  LICENSES,
  USER_DATA,
  SEARCH_DRAFTS,
  DRAFT_HISTORY,
  DRAFT_SEARCH_TAGS,
} from "../../queryKeys";

export interface UseDraft {
  id: number;
  language?: string;
  responsibleId?: string;
}

export interface UseDraftHistory {
  id: number;
  language?: string;
}

export const draftQueryKeys = {
  draft: (params?: Partial<UseDraft>) => [DRAFT, params] as const,
  draftHistory: (params?: Partial<UseDraftHistory>) => [DRAFT_HISTORY, params] as const,
  search: (params?: Partial<DraftSearchQuery>) => [SEARCH_DRAFTS, params] as const,
  licenses: [LICENSES] as const,
  userData: [USER_DATA] as const,
  statusStateMachine: (params?: Partial<StatusStateMachineParams>) => [DRAFT_STATUS_STATE_MACHINE, params] as const,
  draftSearchTags: (params?: Partial<UseSearchTags>) => [DRAFT_SEARCH_TAGS, params] as const,
};

draftQueryKeys.draft({ id: 1 });

export const useDraft = (params: UseDraft, options?: Partial<UseQueryOptions<IArticle>>) => {
  return useQuery<IArticle>({
    queryKey: draftQueryKeys.draft(params),
    queryFn: () => fetchDraft(params.id, params.language),
    ...options,
  });
};
interface UseSearchDrafts extends DraftSearchQuery {
  ids: number[];
}

export const useDraftHistory = (params: UseDraftHistory, options?: Partial<UseQueryOptions<IArticle[]>>) => {
  return useQuery<IArticle[]>({
    queryKey: draftQueryKeys.draft(params),
    queryFn: () => fetchDraftHistory(params.id, params.language),
    ...options,
  });
};

export const useSearchDrafts = (params: UseSearchDrafts, options?: Partial<UseQueryOptions<ISearchResult>>) => {
  return useQuery<ISearchResult>({
    queryKey: draftQueryKeys.search(params),
    queryFn: () => searchAllDrafts(params.ids, params.language, params.sort),
    ...options,
  });
};

export const useLicenses = <ReturnType = ILicense[]>(
  options?: Partial<UseQueryOptions<ILicense[], unknown, ReturnType>>,
) =>
  useQuery<ILicense[], unknown, ReturnType>({
    queryKey: draftQueryKeys.licenses,
    queryFn: fetchLicenses,
    refetchOnWindowFocus: false,
    refetchOnMount: false,
    refetchOnReconnect: false,
    staleTime: Infinity,
    retry: false,
    ...options,
  });

export const useUserData = (options?: Partial<UseQueryOptions<IUserData | undefined>>) =>
  useQuery<IUserData | undefined>({
    queryKey: draftQueryKeys.userData,
    queryFn: fetchUserData,
    ...options,
  });

export const useUpdateUserDataMutation = () => {
  const queryClient = useQueryClient();
  return useMutation<IUserData, unknown, IUpdatedUserData, IUserData>({
    mutationFn: (data) => updateUserData(data),
    onMutate: async (newUserData) => {
      const key = draftQueryKeys.userData;
      await queryClient.cancelQueries({ queryKey: key });
      const previousUserData = queryClient.getQueryData<IUserData>(key);
      queryClient.setQueryData<IUserData>(key, {
        ...previousUserData,
        userId: previousUserData?.userId!,
        ...newUserData,
      });
      return previousUserData;
    },
    onError: (_, __, previousUserData) => {
      if (previousUserData) {
        queryClient.setQueryData(draftQueryKeys.userData, previousUserData);
      }
    },
    onSettled: () => queryClient.invalidateQueries({ queryKey: draftQueryKeys.userData }),
  });
};

interface StatusStateMachineParams {
  articleId?: number;
}

export const useDraftStatusStateMachine = (
  params: StatusStateMachineParams = {},
  options?: Partial<UseQueryOptions<DraftStatusStateMachineType>>,
) => {
  return useQuery<DraftStatusStateMachineType>({
    queryKey: draftQueryKeys.statusStateMachine(params),
    queryFn: () => fetchStatusStateMachine(params.articleId),
    ...options,
  });
};

export interface UseSearchTags {
  input: string;
  language: string;
}

export const useDraftSearchTags = (params: UseSearchTags, options?: Partial<UseQueryOptions<ITagsSearchResult>>) => {
  return useQuery<ITagsSearchResult>({
    queryKey: draftQueryKeys.draftSearchTags(params),
    queryFn: () => fetchSearchTags(params.input, params.language),
    ...options,
  });
};
