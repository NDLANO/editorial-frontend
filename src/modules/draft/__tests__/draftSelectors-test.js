/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { getDraft, getSaving } from '../draft';
import { topicArticleNB, topicArticleEN } from './mockDrafts';

const state = {
  locale: 'nb',
  drafts: {
    isSaving: true,
    all: {
      [topicArticleNB.id]: topicArticleNB,
      2: {
        id: '2',
        created: '2014-12-24T10:44:06Z',
        title: { title: 'Tester', language: 'nb' },
        metaDescription: { metaDescription: 'Beskrivelse', language: 'nb' },
      },
      3: {
        id: '3',
        created: '2014-11-24T10:44:06Z',
        title: { title: 'Tester', language: 'nb' },
      },
      [topicArticleEN.id]: topicArticleEN,
    },
  },
};

test('articleSelectors getArticle with id', () => {
  expect(getDraft(1)(state).id).toBe('1');
  expect(getDraft(2)(state).id).toBe('2');
  expect(getDraft(3)(state).id).toBe('3');
});

test('articleSelectors getArticle (nb locale)', () => {
  const getArticleSelector = getDraft(1);
  expect(getArticleSelector(state)).toMatchSnapshot();
});

test('articleSelectors getArticle (en locale)', () => {
  const getArticleSelector = getDraft('4');
  const stateWithEnLocale = { ...state, locale: 'en' };
  expect(getArticleSelector(stateWithEnLocale)).toMatchSnapshot();
});

test('articleSelectors getArticle returns undefined if article is not in state', () => {
  const getArticleSelector = getDraft(1337);
  expect(getArticleSelector(state)).toEqual(undefined);
});

test('articleSelectors getSaving', () => {
  expect(getSaving(state)).toBe(true);
});
