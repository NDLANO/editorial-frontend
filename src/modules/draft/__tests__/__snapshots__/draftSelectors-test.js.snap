// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`articleSelectors getArticle (en locale) 1`] = `
Object {
  "articleType": "topic-article",
  "content": "<section><p>Tim Berners Lees the World Wide Web is...</p></section>",
  "copyright": Object {
    "creators": Array [
      Object {
        "name": "Ragna Marie Tørdal",
        "type": "writer",
      },
    ],
    "license": Object {
      "description": "Creative Commons Attribution-ShareAlike 2.0 Generic",
      "license": "by-sa",
      "url": "https://creativecommons.org/licenses/by-sa/2.0/",
    },
    "origin": "",
  },
  "created": "2016-11-26T11:07:25Z",
  "footnotes": undefined,
  "id": "4",
  "introduction": "Brainstorming is an important",
  "language": "en",
  "metaDescription": "Brainstorming",
  "oldNdlaUrl": "//ndla.no/node/172732",
  "requiredLibraries": Array [],
  "revision": 1,
  "tags": Array [
    "storytelling",
    "idea development",
    "media design",
    "instruments",
  ],
  "title": "Brainstorming",
  "updated": "2016-12-01T13:05:02Z",
  "updatedBy": "content-import-client",
  "visualElement": "<embed/>",
}
`;

exports[`articleSelectors getArticle (nb locale) 1`] = `
Object {
  "articleType": "topic-article",
  "content": "<section><p>Trykkpressen til Gutenberg og Tim Berners Lees the World Wide Web er begge...</p></section>",
  "copyright": Object {
    "creators": Array [
      Object {
        "name": "Ragna Marie Tørdal",
        "type": "writer",
      },
    ],
    "license": Object {
      "description": "Creative Commons Attribution-ShareAlike 2.0 Generic",
      "license": "by-sa",
      "url": "https://creativecommons.org/licenses/by-sa/2.0/",
    },
    "origin": "",
  },
  "created": "2016-11-26T11:07:25Z",
  "footnotes": undefined,
  "id": "1",
  "introduction": "Idéutvikling er en viktig del av",
  "language": "nb",
  "metaDescription": "Presentasjon av emne om idéutvikling",
  "oldNdlaUrl": "//ndla.no/node/172732",
  "requiredLibraries": Array [],
  "revision": 1,
  "tags": Array [
    "fortelleteknikk",
    "idéutvikling",
    "mediedesign",
    "virkemidler",
  ],
  "title": "Idéskaping og mediedesign",
  "updated": "2016-12-01T13:05:02Z",
  "updatedBy": "content-import-client",
  "visualElement": "<embed/>",
}
`;
